File: backend/routes/notifications.py
from fastapi import APIRouter
from datetime import datetime
from database import db

router = APIRouter(prefix="/api", tags=["notifications"])


@router.get("/notifications")
def get_notifications():
    today = datetime.now().strftime("%Y-%m-%d")
    waste = db.filter("waste_summary", date=today)
    foods = db.get_all("food_items")
    notifications = []

    # High waste alerts
    for w in waste:
        served = w.get("total_served_kg") or w.get("served_kg") or 0
        wasted = w.get("total_wasted_kg") or w.get("wasted_kg") or 0
        pct = (wasted / served * 100) if served > 0 else 0
        if pct > 60:
            food = next((f for f in foods if f["id"] == (w.get("item_id") or w.get("food_item_id"))), None)
            notifications.append({
                "id": f"waste_{w.get('id')}", "type": "high_waste", "severity": "critical",
                "title": f"⚠️ High waste: {food['name'] if food else 'Unknown'}",
                "message": f"{food['name'] if food else 'Unknown'} waste is {pct:.0f}% during {w.get('meal_type')}",
                "timestamp": w.get("calculated_at") or datetime.now().isoformat(),
                "read": False,
            })

    # Low review alerts
    reviews = db.get_all("reviews")
    for r in reviews:
        if r.get("date") == today and r.get("overall_rating", 5) <= 2:
            notifications.append({
                "id": f"review_{r.get('id')}", "type": "low_rating", "severity": "warning",
                "title": f"⭐ Low rating for {r.get('meal_type')}",
                "message": f"{r.get('student_name')} rated {r.get('meal_type')} {r.get('overall_rating')}★: \"{r.get('comment')}\"",
                "timestamp": r.get("created_at") or datetime.now().isoformat(),
                "read": False,
            })

    # Camera status
    notifications.append({
        "id": "camera_status", "type": "system", "severity": "info",
        "title": "📷 All cameras online",
        "message": "3/3 cameras are active and processing",
        "timestamp": datetime.now().isoformat(),
        "read": True,
    })

    notifications.sort(key=lambda n: n.get("timestamp", ""), reverse=True)
    return {"success": True, "data": notifications}

File: routes/notifications.ts
// Notification Routes
import { Express } from "express";
import { Server } from "socket.io";

export function registerNotificationRoutes(app: Express, db: any, io: Server) {
  // Get notifications (simulated from data)
  app.get("/api/notifications", (req, res) => {
    try {
      const today = new Date().toISOString().split('T')[0];
      const waste = db.prepare("SELECT * FROM waste_summary WHERE date = ?").all(today);
      const foods = db.prepare("SELECT * FROM food_items").all();
      const notifications: any[] = [];

      // High waste alerts
      waste.forEach((w: any) => {
        const pct = (w.total_served_kg || w.served_kg) > 0
          ? ((w.total_wasted_kg || w.wasted_kg) / (w.total_served_kg || w.served_kg)) * 100 : 0;
        if (pct > 60) {
          const food = foods.find((f: any) => f.id === (w.item_id || w.food_item_id));
          notifications.push({
            id: `waste_${w.id}`, type: 'high_waste', severity: 'critical',
            title: `⚠️ High waste: ${food?.name || 'Unknown'}`,
            message: `${food?.name} waste is ${pct.toFixed(0)}% during ${w.meal_type}`,
            timestamp: w.calculated_at || new Date().toISOString(), read: false
          });
        }
      });

      // Low review alerts
      const reviews = db.prepare("SELECT * FROM reviews ORDER BY id DESC").all();
      reviews.filter((r: any) => r.date === today && r.overall_rating <= 2).forEach((r: any) => {
        notifications.push({
          id: `review_${r.id}`, type: 'low_rating', severity: 'warning',
          title: `⭐ Low rating for ${r.meal_type}`,
          message: `${r.student_name} rated ${r.meal_type} ${r.overall_rating}★: "${r.comment}"`,
          timestamp: r.created_at || new Date().toISOString(), read: false
        });
      });

      // Camera status (always online for demo)
      notifications.push({
        id: 'camera_status', type: 'system', severity: 'info',
        title: '📷 All cameras online',
        message: '3/3 cameras are active and processing',
        timestamp: new Date().toISOString(), read: true
      });

      res.json({ success: true, data: notifications.sort((a, b) => b.timestamp.localeCompare(a.timestamp)) });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });
}
